### PR TITLE
SNMPv3 support

### PR DESCRIPTION
**Pull request by colinsurprenant**
This PR ads support for SNMPv3 and replaces #21 by @axrayn. I took all original commits from #21 and added new commits to continue on the original work. I essentially refactored the `Client` and `ClientV3` classes using a `BaseClient` class to DRY code. 

Also @axrayn originally offered the option to specify multiple SNMPv3 users but I decided to revert that and now only a single SNMPv3 user can be configured for all defined hosts. If multiple SNMPv3 users are required, for now, this will have to be done using multiple snmp input definitions. This was done so that the password option fields have the `password` validation to avoid leaking plaintext password in logs for example. We will see how we will address this limitation in the future.

I tested SNMPv3 support manually on my local machine. 

**Reviewers**: Unfortunately I haven't added a tests SNMP server to be able to add specs for the actual SNMP calls so currently tests must be made against an external SNMP server/device. I am personally using net-snmp on OSX for this purpose. I manually tested SNMPv3 support and will rely on the beta cycle to have users test it in other contexts. Manual tests can be made using the `test/test_client.rb` script or using logstash and use the sample config `test/sample.conf`.

### CODE REVIEW

**Comment by axrayn** [comment]
I'll update later on tonight and test it out.

Moving back to a single v3 user per input works too. We have roughly 8 different v3 users to work with but that's generally separated across different hw types so distinct input would probably work out better for us in the long run.

**Comment by colinsurprenant** [comment]
@axrayn yeah ... I'll think about the way we could have multiple v3 users, as I said that choice to revert to a single one for now is really motivated by the `password` option handling - we really do have to be cautious  with passwords handling and the potential to leak them in logs and such. I will review what was done in the [azure_event_hubs](https://github.com/logstash-plugins/logstash-input-azure_event_hubs) by @jakelandis with the basic & advanced modes of configuration and we can probably reuse this strategy here, but this will be post-GA.

File: lib/logstash/inputs/snmp.rb [hunk 1]
```diff
@@ -189,6 +220,15 @@ def validate_oids!
     raise(LogStash::ConfigurationError, "at least one get OID or one walk OID is required") if @get.empty? && @walk.empty?
   end
 
+  def validate_v3_user!
+    raise(LogStash::ConfigurationError, "v3 user must have a \"security_name\" option") if @security_name.nil?
+    raise(LogStash::ConfigurationError, "you must specify an auth protocol if you specify an auth pass") if  @auth_protocol.nil? && !@auth_pass.nil?
+    raise(LogStash::ConfigurationError, "you must specify an auth pass if you specify an auth protocol") if !@auth_protocol.nil? &&  @auth_pass.nil?
+    raise(LogStash::ConfigurationError, "you must specify a priv protocol if you specify a priv pass")   if  @priv_protocol.nil? && !@priv_pass.nil?
+    raise(LogStash::ConfigurationError, "you must specify a priv pass if you specify a priv protocol")   if !@priv_protocol.nil? &&  @priv_pass.nil?
+   end
```
Comment (guyboertje) [inline lib/logstash/inputs/snmp.rb]: nitpick (resolve if you disagree): 
As I make mistakes and typos a lot I would like to see all my mistakes in one error message so I can fix it once and not fix the error presented to me and then see another one.
Will you consider pushing the error string to an array when the validation condition fails and raise a joined message once when the array is not empty?

Comment (colinsurprenant) [reply lib/logstash/inputs/snmp.rb]: good suggestion!

**Comment by colinsurprenant** [comment]
Thanks @guyboertje. Will go ahead and merge, and will squash by authors to not loose authorship information.
@axrayn I will be publishing a beta5 shortly so we can have more testers for this. Don't hesitate to comment here or in a new issue for any problems you see.
Again, thanks a lot for you contribution on this @axrayn !

**Comment by colinsurprenant** [comment]
included in the just released 0.1.0.beta5

**Pull request merged**